proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,200 @@
+
+A multi-signature authority, comprised of key-holders from Solana Labs and
+possibly other validator client teams in the future, will have the authority to
+stage created features for activation.
+In the future, this authority could be replaced by validator governance.
+
+The multi-signature authority stages a feature for activation by invoking the
+Feature Gate program's `StageFeatureForActivation` instruction. This instruction
+expects:
+
+- Data: The feature ID
+- Accounts:
+  - Staged Features PDA: writable
+  - Multi-signature authority: signer 
+
+`StageFeatureForActivation` will add the provided feature ID to the **next
+epoch's** Staged Features PDA.
+
+The Staged Features PDA for a given epoch stores a list of all feature IDs that
+were staged for activation during the previous epoch. This list shall have a
+maximum length of 8 (ie. `[Pubkey; 8]`).
+
+The proposed seeds for deriving the Staged Features PDA are provided below,
+where the `<epoch>` is the epoch during which the contained feature IDs will be
+assessed for stake support and considered for activation.
+
+```
+"staged_features" + <epoch>
+```
+
+### Step 3: Signaling Support for Staged Features
+
+With an on-chain reference point to determine the features staged for activation
+for a particular epoch, nodes can signal their support for the staged features
+supported by their software.
+
+A node signals its support for staged features by invoking the Feature Gate
+program's `SignalSupportForStagedFeatures` instruction. This instruction
+expects:
+
+- Data: A `u8` bit mask of the staged features.
+- Accounts:
+  - Support Signal PDA: writable
+  - Vote account: signer
+
+A `1` bit represents support for a feature. For example, for staged features
+`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
+except `E` and `H`, their `u8` value would be 246, or `11110110`.
+
+A node's submitted bit mask is then stored in a Support Signal PDA derived from
+that node's vote address. The proposed seeds are defined below.
+
+```
+"support_signal" + <vote_address>
+```
+
+Nodes should send a transaction containing this instruction at some arbitrary
+point during the epoch at least 128 slots before the end of the epoch and on
+startup after any reboot.
+
+Note: If a feature is revoked, the list of staged features will not change, and
+nodes may still signal support for this feature. However, the runtime will not
+activate this feature if its corresponding feature account no longer exists
+on-chain.
+
+### Step 4: Activation and Garbage Collection
+
+During the epoch rollover, the runtime uses the validator support signals to
+determine which staged features to activate.
+
+To do this, the runtime walks all of the vote accounts, derives their Support
+Signal PDA to read their bit mask, and tallies up the total stake support for
+each staged feature. The runtime will also zero-out each bit mask, resetting
+each Support Signal PDA for the next epoch.
```

Comment:
I think it would be safer to put as much of the logic on-chain as possible, to minimise the chance of a partition.

Instead of the runtime walking the vote accounts, the tally could be updated on-chain incrementally, in an account owned by the feature program. Then at the start of each epoch, the runtime can just look up the thresholds from this account.

Reply:
This is interesting, except the only issue is that the Feature Gate program has no way to know who _didn't_ vote, so there might still be a walking step in the runtime anyway.

Reply:
If we expose `StakeHistory` to BPF programs, which we will likely have to do as part of the migration, the Feature Gate program can use that to determine which proportion of the stake has voted.

Reply:
Alternatively, if we can't expose it, we can see about this approach:
https://github.com/solana-labs/solana-program-library/pull/6106

Reply:
Or https://github.com/solana-labs/solana/pull/35252

Reply:
@topointon-jump , how are you envisioning the program would calculate the specific stake amount/percentage of a vote account? The `StakeHistory` sysvar only contains the stake totals for the epoch (effective, activating, deactivating), and a vote account doesn't know how much stake is delegated to it.
I haven't thought of a way to move that tally on-chain without a massive undertaking.

Reply:
@CriesofCarrots just for my understanding, why do we need the specific stake amount/percentage of a vote account? I thought all we needed to do was work out what proportion of the current stake for an epoch had sent support signals. We can use StakeHistory to tell us how much stake is activated for the epoch, and the Feature Gate program can update the on-chain tally as support signals are received, looking up each vote accounts delegated stake from the account values (not sysvars) as each support signal is received. 

Let me know if I'm not understanding something, also happy to hop on a call anytime :)

Reply:
> looking up each vote accounts delegated stake from the account values

As I said, a vote account doesn't know how much stake is delegated to it, so the only way currently to do what you describe would be to pass every delegating stake account into the support transaction. That's obviously not feasible 😅 

Adding a new syscall that does `get_epoch_stake(validator_id: Pubkey) -> u64` as per https://discord.com/channels/428295358100013066/1141064343027462214/1214301700924510238 would enable tallying on a support transaction, however. Just trying to be clear that there is no existing mechanism to do this.

Reply:
> Adding a new syscall that does get_epoch_stake(validator_id: Pubkey) -> u64 as per ... would enable tallying on a support transaction, however.

@CriesofCarrots it sounds like this might be a good route to take in order to reduce complexity on the runtime.

How concerned are you with the O(n) tallying happening on epoch rollover? It seems like if creating this `get_epoch_stake` function isn't too much of a lift, it's definitely a more optimized solution.

Alternatively, there's also the block header route discussed in your linked Discord conversation.

Reply:
Sorry @CriesofCarrots I misunderstood what you meant! Yes, imo adding the `get_epoch_stake` syscall is the easiest way to achieve this, if we go down this route.

Reply:
@ripatel-fd , yes, I think that would be a great next step. Thank you!

Reply:
@ripatel-fd https://github.com/solana-foundation/solana-improvement-documents/pull/133

Reply:
The SIMD for the `getEpochStake` syscall was merged. Perhaps soon we can have another review on this?